lib/foreman_remote_execution_core/script_runner.rb

```diff
@@ -7,30 +7,116 @@
+      end
+    end
+
+    def sent_all_data?
+      effective_user_password.nil? || password_sent
+    end
+
+    def cli_command_prefix
+      "sudo -p '#{LOGIN_PROMPT}' -u #{effective_user} "
+    end
+
+    def reset
+      @password_sent = false
+    end
+  end
+
+  class SuUserMethod
+    attr_accessor :effective_user, :ssh_user
+
+    def initialize(effective_user, ssh_user)
+      @effective_user = effective_user
+      @ssh_user = ssh_user
+    end
+
+    def on_data(_, _); end
+
+    def sent_all_data?
+      true
+    end
+
+    def cli_command_prefix
+      "su - #{effective_user} -c "
+    end
+
+    def reset; end
+  end
+
+  class NoopUserMethod
+    def on_data(_, _); end
+
+    def sent_all_data?
+      true
+    end
+
+    def cli_command_prefix; end
+
+    def reset; end
+  end
+
   class ScriptRunner < ForemanTasksCore::Runner::Base
     attr_reader :execution_timeout_interval
 
     EXPECTED_POWER_ACTION_MESSAGES = ['restart host', 'shutdown host'].freeze
     DEFAULT_REFRESH_INTERVAL = 1
     MAX_PROCESS_RETRIES = 4
 
-    def initialize(options)
+    def initialize(options, user_method)
       super()
       @host = options.fetch(:hostname)
       @script = options.fetch(:script)
       @ssh_user = options.fetch(:ssh_user, 'root')
       @ssh_port = options.fetch(:ssh_port, 22)
       @ssh_password = options.fetch(:secrets, {}).fetch(:ssh_password, nil)
       @key_passphrase = options.fetch(:secrets, {}).fetch(:key_passphrase, nil)
-      @effective_user = options.fetch(:effective_user, nil)
-      @effective_user_method = options.fetch(:effective_user_method, 'sudo')
       @host_public_key = options.fetch(:host_public_key, nil)
       @verify_host = options.fetch(:verify_host, nil)
       @execution_timeout_interval = options.fetch(:execution_timeout_interval, nil)
 
       @client_private_key_file = settings.fetch(:ssh_identity_key_file)
       @local_working_dir = options.fetch(:local_working_dir, settings.fetch(:local_working_dir))
       @remote_working_dir = options.fetch(:remote_working_dir, settings.fetch(:remote_working_dir))
+
+      @user_method = user_method
+    end
+
+    def self.build(options)
+      effective_user = options.fetch(:effective_user, nil)
+      ssh_user = options.fetch(:ssh_user, 'root')
+      effective_user_method = options.fetch(:effective_user_method, 'sudo')
+
+      user_method = if effective_user.nil? || effective_user == ssh_user
+                      NoopUserMethod.new
+                    elsif effective_user_method == 'sudo'
+                      SudoUserMethod.new(options.fetch(:effective_user, nil),
+                                         options.fetch(:ssh_user, 'root'),
+                                         options.fetch(:secrets, {}).fetch(:sudo_password, nil))
+                    elsif effective_user_method == 'su'
+                      SuUserMethod.new(options.fetch(:effective_user, nil),
+                                       options.fetch(:ssh_user, 'root'))
```

Comment:
Why not reuse the `effective_user` and `ssh_user` extracted on lines 102 and 103?